pkg/cache/json.go

```diff
@@ -230,56 +234,141 @@ func (q *JSON) Build(ctx context.Context, fbcFsys fs.FS) error {
 		return fmt.Errorf("ensure clean base directory: %v", err)
 	}
 
-	fbc, err := declcfg.LoadFS(ctx, fbcFsys)
+	tmpFile, err := os.CreateTemp("", "opm-cache-build-")
 	if err != nil {
-		return err
+		return fmt.Errorf("create temp file: %v", err)
 	}
-	fbcModel, err := declcfg.ConvertToModel(*fbc)
-	if err != nil {
-		return err
+	defer func() {
+		_ = tmpFile.Close()
+		_ = os.Remove(tmpFile.Name())
+	}()
+
+	packageReaders := map[string][]io.Reader{}
+	walkMetasMu := sync.Mutex{}
+	offset := int64(0)
+	if err := declcfg.WalkMetasFS(ctx, fbcFsys, func(path string, meta *declcfg.Meta, err error) error {
+		if err != nil {
+			return err
+		}
+		walkMetasMu.Lock()
+		defer walkMetasMu.Unlock()
+		if _, err := tmpFile.Write(meta.Blob); err != nil {
+			return fmt.Errorf("failed to write blob from %q: %v", path, err)
+		}
+		blobLen := int64(len(meta.Blob))
+		sr := io.NewSectionReader(tmpFile, offset, blobLen)
+		offset += blobLen
+
+		if meta.Schema == declcfg.SchemaPackage {
+			packageReaders[meta.Name] = append(packageReaders[meta.Name], sr)
+		} else if meta.Package != "" {
+			packageReaders[meta.Package] = append(packageReaders[meta.Package], sr)
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("failed to walk metas: %v", err)
 	}
 
-	pkgs, err := packagesFromModel(fbcModel)
-	if err != nil {
+	q.apiBundles = map[apiBundleKey]string{}
+	q.packageIndex = packageIndex{}
+	pkgChan := make(chan string, runtime.NumCPU())
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	eg.Go(func() error {
+		defer close(pkgChan)
+		for pkgName := range packageReaders {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case pkgChan <- pkgName:
+			}
+		}
+		return nil
+	})
+
+	var mapMu sync.Mutex
+	for i := 0; i < runtime.NumCPU(); i++ {
```

Comment:
This means that data for up to `runtime.NumCPU` packages will be in memory at once. I noticed the following when building a cache for the operatorhub catalog on my 10-core Mac M1 Pro:

On `master`:
- With unmigrated catalog
	- With `GOMAXPROCS=1`: Peak memory=778Mb, Duration=15s
	- With unset GOMAXPROCS, on master branch: Peak memory=731Mb, Duration=10s
- With migrated catalog
	- With `GOMAXPROCS=1` on master branch: Peak memory=166Mb, Duration=2.7s
	- With unset GOMAXPROCS, on master branch: Peak memory=141Mb, Duration=2.0s

On PR branch:
- With unmigrated catalog
	- With `GOMAXPROCS=1` on PR branch: Peak memory=234Mb, Duration=16s
	- With unset GOMAXPROCS, on PR branch: Peak memory=290Mb, Duration=6.6s
- With migrated catalog
	- With `GOMAXPROCS=1` on PR branch: Peak memory=115Mb, Duration=3s
	- With unset GOMAXPROCS, on PR branch: Peak memory=117Mb, Duration=1.39s